storage/replica.go

```diff
@@ -1229,67 +1405,68 @@ func (r *Replica) resolveIntents(ctx context.Context, intents []proto.Intent) {
-				// than the pushee's txn!
-				Timestamp: intent.Txn.Timestamp,
-				Key:       intent.Key,
-				EndKey:    intent.EndKey,
-				Txn:       &intent.Txn,
+				Key:    intent.Key,
+				EndKey: intent.EndKey,
 			}
 
 			if len(intent.EndKey) == 0 {
-				resolveArgs = &proto.ResolveIntentRequest{RequestHeader: header}
+				resolveArgs = &proto.ResolveIntentRequest{
+					RequestHeader: header,
+					IntentTxn:     intent.Txn,
+				}
 				local = r.ContainsKey(intent.Key)
 			} else {
-				resolveArgs = &proto.ResolveIntentRangeRequest{RequestHeader: header}
+				resolveArgs = &proto.ResolveIntentRangeRequest{
+					RequestHeader: header,
+					IntentTxn:     intent.Txn,
+				}
 				local = r.ContainsKeyRange(intent.Key, intent.EndKey)
 			}
 		}
 
 		// If the intent isn't (completely) local, we'll need to send an external request.
 		// We'll batch them all up and send at the end.
-		if !local {
+		if local {
+			bArgsLocal.Add(resolveArgs)
+		} else {
 			bArgs.Add(resolveArgs)
-			continue
 		}
+	}
 
-		// If it is local, it goes directly into Raft.
-		// TODO(tschottdorf): this may be premature optimization. Consider just
-		// treating everything as an external request. This means having to
-		// wait for complete execution of the command (whereas now we just wait
-		// for proposition) and some more overhead sending things around.
-		wg.Add(1)
-		action := func() {
-			// Trace this under the ID of the intent owner.
-			ctx := tracer.ToCtx(ctx, r.rm.Tracer().NewTrace(resolveArgs.Header().Txn))
-			if _, err := r.addWriteCmd(ctx, resolveArgs, &wg); err != nil && log.V(1) {
-				log.Warningc(ctx, "resolve for key %s failed: %s", intent.Key, err)
-			}
+	// The local batch goes directly to Raft.
+	var wg sync.WaitGroup
+	action := func() {
+		// Trace this under the ID of the intent owner.
+		ctx := tracer.ToCtx(ctx, r.rm.Tracer().NewTrace(bArgsLocal.Header()))
+		if _, err := r.addWriteCmd(ctx, bArgsLocal, &wg); err != nil && log.V(1) {
+			log.Warningc(ctx, "batch resolve failed: %s", err)
 		}
+	}
+	if len(bArgsLocal.Requests) > 0 {
+		wg.Add(1)
 		if !r.rm.Stopper().RunAsyncTask(action) {
-			// Still run the task. Our caller already has a task and going async
-			// here again is merely for performance, but some intents need to
-			// be resolved because they might block other tasks. See #1684.
-			// Note that handleSkippedIntents has a TODO in case #1684 comes
-			// back.
+			// Still run the task when draining. Our caller already has a task and
+			// going async here again is merely for performance, but some intents
+			// need to be resolved because they might block other tasks. See #1684.
+			// Note that handleSkippedIntents has a TODO in case #1684 comes back.
 			action()
 		}
 	}
+
 	// Resolve all of the intents which aren't local to the Range. This is a
 	// no-op if all are local.
 	b := &client.Batch{}
 	b.InternalAddCall(proto.Call{Args: bArgs, Reply: &proto.BatchResponse{}})
-	action := func() {
```

Comment:
ditto: declare this only if needed by putting it inside the `if len(bArgs.Requests) > 0`